SW-653 mr beam to use update config from cloud

JFSolution opened the pull request:
this feature enables us to use the update config from https://github.com/mrbeam/beamos_config/blob/master/docs/sw-update-conf.json to be more flexible in changing the config

Commit:
changed update config to a file and load this from a server
https://github.com/mrbeam/issues/issues/60
https://github.com/mrbeam/issues/issues/59

Commit:
refactor

Commit:
changed local file path to octoprint folder
changed file version system to int
merged changes from buster branch
added beamos version check
added some errorhandling
fixed override order
https://github.com/mrbeam/issues/issues/59

Commit:
fixed config for octoprint
https://github.com/mrbeam/issues/issues/59

Commit:
added threading for the loading of the file from the server
added some errorhandling
https://github.com/mrbeam/issues/issues/59

Commit:
added/updatet the documentation
https://github.com/mrbeam/issues/issues/59

Commit:
added debug logging for pip command
https://github.com/mrbeam/issues/issues/59

Commit:
changed url to hosted file
https://mr-beam.atlassian.net/browse/SW-88

Commit:
SW-80 added fallbackfile and fallbackfile handling
included all requested PR changes

Commit:
SW-80 only override file if a new version is available or local file is not valid

Commit:
SW-80 moved modules in json file in module section
changed threading concept
refactor the update file path

Commit:
SW-653 refactoring, added some tests

amrsoll commented on octoprint_mrbeam/software_update_information.py:
```diff
@@ -19,72 +24,197 @@
 
+_softwareupdate_handler = None
+
+class MrBeamSoftwareupdateHandler:
+    CLOUD_FILE = 1
+    LOCAL_FILE = 2
+    REPO_FILE = 3
+    def __init__(self, plugin):
+        self._plugin = plugin
```
1. You do not seem to make the best use of the class structure. Are you sure it is the best method to implement the feature?

2. It does not seem you need the `plugin` object, but rather a subset of information (the selected SW channel). Do you foresee any scenario where you need to use the plugin item itself?

JFSolution replied:
1. I will check, I think there was something why I choose the class structure
2. plugin will be used to get the path to the local saved file

amrsoll replied:
sure, but then you could just receive the path if it's the only info you need, not the whole plugin :)

JFSolution replied:
is also used to get the versionnumber of the octoprint plugins, so I will keep the plugin

amrsoll commented on octoprint_mrbeam/software_update_information.py:
```diff
@@ -19,72 +24,197 @@
+                "fallback use local default config file /files/software_update/update_info.json"
+            )
+            try:
+                copy(
```
please write `shutil.copy` as it is less confusing (`copy()` is usually an object copy, from the `copy` package.)

JFSolution replied:
ok

amrsoll commented on octoprint_mrbeam/software_update_information.py:
```diff
@@ -19,72 +24,197 @@
+            sw_update_plugin._version_cache_dirty = True
+        return returncode
+
+@logme(output=True)
```
Do you still need it?

JFSolution replied:
no the logme can be removed before the merge, I only needed it for debuging

amrsoll commented on octoprint_mrbeam/software_update_information.py:
```diff
@@ -19,72 +24,197 @@
+
+    # @get_thread(daemon=False)
+    @logExceptions
+    def load_update_file_from_cloud(self, localfilemissing=False):
```
What is the benefit of the `localfilemissing` variable ? Because the function checks for the file anyway, so it might as well take care of it itself.

I also think this function is trying to do too many things at the same time: downloading a file and managing local copies.

I would recommend to have different functions that
- tries to download the file and can raise an exception
- verify the presence and content of a local file path
- a function that uses the 2 previous ones to dictate the usage flow : if download fails, use the local copy etc...

This will also help you for the testing so variables like `localfilemissing` are more natural and you have more flexibility.

JFSolution replied:
This will be used to force to ignore the local file and use the server file, ok I seperate them

amrsoll commented on octoprint_mrbeam/software_update_information.py:
```diff
@@ -19,72 +24,197 @@
+
 BEAMOS_LEGACY_DATE = date(2018, 1, 12)
 
+_softwareupdate_handler = None
```
If you are going to use a singleton, then please explain (docstring) this variable

JFSolution replied:
ok

amrsoll commented on octoprint_mrbeam/software_update_information.py:
```diff
@@ -19,72 +24,197 @@
-            )
-        )
-    return res
+    _logger.info("SoftwareUpdate using tier: %s %s", tier, beamos_date)
+    global _softwareupdate_handler
+    if _softwareupdate_handler is None:
+        _softwareupdate_handler = MrBeamSoftwareupdateHandler(plugin)
+        _softwareupdate_handler.load_update_file_from_cloud()
+
+    return _set_info_from_file(plugin, tier, beamos_date, _softwareupdate_handler)
```
please do not use singletons. You can just create a new instance, there is no benefit to have a singleton here.

JFSolution replied:
but then it can run multiple processes that will fetch the file from the server, this will make sure that only one process is running that checks the file

amrsoll replied:
Please use a lock if you want to manage threaded access.

JFSolution replied:
is it done correlty now?

amrsoll commented on octoprint_mrbeam/software_update_information.py:
```diff
@@ -98,233 +228,190 @@ def switch_software_channel(plugin, channel):
-    pluginInfo = plugin._plugin_manager.get_plugin_info(module_id)
-    if pluginInfo is None:
-        display_version = None
+def _set_info_from_file(plugin, tier, beamos_date, _softwareupdate_handler):
+    """
+    loads update info from the update_info.json file
+    the override order: default_settings->module_settings->tier_settings->beamos_settings
+    and if there are update_settings set in the config.yaml they will replace all of the module
+    the json file should look like:
```
I think the name is misleading. It sounds like a function that would take a `file` or a `path` variable, and simply parse the content.

I would recommend to have a parser function separately and some other function that takes care of providing that file, use the parser and return the result to the OctoPrint hook.

JFSolution replied:
I don't see why i should seperate this, it will only open the file catch some exceptions and the parse the content

amrsoll replied:
Please use a(n) `(R)Lock` if you want to prevent simultaneous file access. 

Using singletons is more of a coding style for property managers, which I don't think is necessarily good. They carry a baggage of globalised side effects and their influence on the behaviour of other components is difficult to test because they are often coded to be mutable like the settings manager.

JFSolution replied:
for this methode is no singelton used

JFSolution replied:
as discussed in the sync meeting, i will use the same lock as for the loading of the file

amrsoll commented on octoprint_mrbeam/software_update_information.py:
```diff
@@ -98,233 +228,190 @@ def switch_software_channel(plugin, channel):
+    pop_list = ["alpha", "beta", "stable", "develop", "beamos_date", "name"]
+    for pop_element in pop_list:
+        if pop_element in update_config:
+            update_config.pop(pop_element)
```
you could also use the `del` keyword.
```python
del update_config[pop_element]
```

JFSolution replied:
this is not working

amrsoll replied:
sorry my bad the `del` keyword confuses the python list comprehension. I corrected my comment.

`del` is more specific for removing pointers.

amrsoll replied:
but you can also make it simpler with
```python
for key in set(update_config).intersection(pop_list):
  del update_config[key]
```

amrsoll commented on setup.py:
```diff
@@ -48,6 +48,7 @@
     "pillow",
     "lxml",
     "numpy",
+    "yaml",
```
from now on we should try to freeze the dependencies if you could

amrsoll commented on tests/softwareupdate/test_cloud_config.py:
```diff
@@ -0,0 +1,219 @@
+    def get_plugin_info(self, module_id):
+        return self
+
+    # sw_update_plugin = self._plugin._plugin_manager.get_plugin_info(
+    #     "softwareupdate"
+    # ).implementation
+    # sw_update_plugin._refresh_configured_checks = True
+    # sw_update_plugin._version_cache = dict()
+    # sw_update_plugin._version_cache_dirty = True
```
raus bitte :)

JFSolution replied:
done

amrsoll reviewed (commented):
Why is the update config in the code? We can preload it in the BeamOS filesystem if we want to have it on the devices that ship offline. And we can add it to the usb sticks in the update module. I don't think it is the best idea to hardcode this config.

I cannot see where you parse the "dependencies" key from the update config file, did I miss it? Or we don't use it yet?

Commit:
freeze versions of dependencies

Commit:
reformat files

Commit:
add requested changes

Commit:
remove singelton

Commit:
added lock for reading the config

Commit:
changed cloud config to a github tag/release based file

Commit:
fix pyyaml version

Commit:
hides the force update button, change behaviour of the check for update button to reload cloud file

Commit:
don't hide "no update info" message

Commit:
add migration to enable the octoprint online check

Commit:
add a host change for the onlinecheck to the migration

Commit:
change retrys and backoff time (time between retrys)
removed unused stuff

Commit:
added restart functionality for migration
do a restart of octoprint after migration

Commit:
added abstraction layer for octoprint onlinechecker

Commit:
version bump

Commit:
changed ip for online check to find.mr-beam.org ip

Commit:
config will be generated for submodules during load
added more testcases and simplified the testing

Commit:
remove debug logging

Yelko-mrbeam commented on octoprint_mrbeam/software_update_information.py:
```diff
@@ -32,59 +43,188 @@
+        config_tag = get_tag_of_github_repo("beamos_config")
+        # if plugin._connectivity_checker.check_immediately():  # check if device online
+        if config_tag:
+            cloud_config = get_config_of_tag(get_tag_of_github_repo("beamos_config"))
```
We could directly use config_tag right? Instead of calling get_tag_of_github_repo again

JFSolution replied:
good point

Yelko-mrbeam commented on octoprint_mrbeam/software_update_information.py:
```diff
@@ -32,59 +43,188 @@
+        else:
+            _logger.warning("no valid response for the tag of the update_config file")
+            return None
+    except requests.ReadTimeout:
```
is this thrown after the 3rd retry if there is another timeout? I would have expected some RetryException 🤔

Yelko-mrbeam commented on octoprint_mrbeam/software_update_information.py:
```diff
@@ -32,59 +43,188 @@
+        url = (
+            "https://api.github.com/repos/mrbeam/beamos_config/contents/docs/sw-update-conf.json"
+            + "?ref=v"
+            + str(tag)
```
what does this "ref" query param imply?

JFSolution replied:
with ref you can define the tag of which you want the file from

Yelko-mrbeam replied:
Ah true.. now that you mention it it seems obvious 😄

JFSolution committed:
Delete update_info file as it is no more needed

Yelko-mrbeam commented on octoprint_mrbeam/__init__.py:
```diff
@@ -966,6 +973,11 @@ def on_wizard_finish(self, handled):
 
     # map(lambda m: m(handled), self._get_subwizard_attrs("_on_", "_wizard_finish").values())
 
+    @octoprint.plugin.BlueprintPlugin.route("/fetch_update_info", methods=["GET"])
```
I would encourage creating a new file like update_handler.py under a package called rest_handlers like I did on this [PR](https://github.com/mrbeam/MrBeamPlugin/pull/1415/files). This would allow us to gather all update endpoint related functionality there. Right now it's only one endpoint but it might grow...

Also I would try to follow a more domain driven pattern to make it easier to add functionality like .route('/update/info'). As far as I understand, this endpoint is actually deleting some cache right? if this is the case I would actually vote for making this a DELETE request instead of a GET one.

JFSolution replied:
I would not change it to DELETE as it executes the check for updates by clearing the version cache

JFSolution replied:
Ok I will move the blueprint as you described

Yelko-mrbeam commented on octoprint_mrbeam/software_update_information.py:
```diff
@@ -98,242 +238,209 @@ def switch_software_channel(plugin, channel):
+    # fmt: on
+    sw_update_plugin._refresh_configured_checks = True
+    sw_update_plugin._version_cache = dict()
+    sw_update_plugin._version_cache_dirty = True
```
It's not great that we need to rely on the private properties... But I guess there is no otherway so far 😿

Yelko-mrbeam commented on octoprint_mrbeam/software_update_information.py:
```diff
@@ -32,59 +43,188 @@
+    sets the config for the Octoprint Softwareupdate Plugin with the data from the config file
+    @param plugin: calling plugin
+    @return: the config for the Octoprint embedded softwareupdate Plugin
+    """
```
Nice doc! Not only here but in all public methods! 👏 ❤️

Yelko-mrbeam commented on octoprint_mrbeam/software_update_information.py:
```diff
@@ -32,59 +43,188 @@
+    ).implementation
+
+    sw_update_plugin._version_cache = dict()
+    sw_update_plugin._version_cache_dirty = True
```
what does this part of the code mean? does it make sense to separate it in a method? it seems similar to the one you use at reload_update_info. We are using private properties and we might be able to change this in the future so it might be a good idea to wrap it in a descriptive method. What do you think?

Yelko-mrbeam commented on octoprint_mrbeam/software_update_information.py:
```diff
@@ -98,242 +238,209 @@ def switch_software_channel(plugin, channel):
+                SW_UPDATE_TIER_BETA,
+                SW_UPDATE_TIER_DEV,
+                SW_UPDATE_TIER_PROD,
+                SW_UPDATE_TIER_ALPHA,
```
Might make sense to make this values an Enum?

Yelko-mrbeam commented on octoprint_mrbeam/software_update_information.py:
```diff
@@ -98,242 +238,209 @@ def switch_software_channel(plugin, channel):
+        defaultsettings = cloud_config.get("default", None)
+        modules = cloud_config["modules"]
+
+        # TODO maybe drop the higher levels of the config so the output will be more tidy
```
What is this TODO about? shall we create a ticket for it and link it to the comment or should we just do it before merging?

Yelko-mrbeam commented on octoprint_mrbeam/software_update_information.py:
```diff
@@ -98,242 +238,209 @@ def switch_software_channel(plugin, channel):
+        if tierversion in input_moduleconfig:
+            input_moduleconfig = dict_merge(
+                input_moduleconfig, input_moduleconfig[tierversion]
+            )  # override tier config from tiers set in config_file
```
this lines are exactly the same as the previous ones except for the comment right? is this correct? 😕

Yelko-mrbeam commented on octoprint_mrbeam/software_update_information.py:
```diff
@@ -98,242 +238,209 @@ def switch_software_channel(plugin, channel):
+                        beamos_config = dict_merge(
+                            beamos_config, beamos_config_module_tier
+                        )  # override tier config from tiers set in config_file
+                    input_moduleconfig = dict_merge(input_moduleconfig, beamos_config)
```
can we extract this to a separate self-explanatory method?

Yelko-mrbeam commented on octoprint_mrbeam/software_update_information.py:
```diff
@@ -98,242 +238,209 @@ def switch_software_channel(plugin, channel):
+                        )  # override tier config from tiers set in config_file
+                    input_moduleconfig = dict_merge(input_moduleconfig, beamos_config)
+
+        if "branch" in input_moduleconfig and "{tier}" in input_moduleconfig["branch"]:
```
I think it makes sense to have a private method for each if on this public method. I think it would make it more readable and we would follow the single responsibility principle. What do you think @Josef-MrBeam ?

If we decide to go for it I would suggest to separate the methods so that they return what you need and you modify the input_moduleconfig in this public method.

Yelko-mrbeam commented on octoprint_mrbeam/software_update_information.py:
```diff
@@ -98,242 +238,209 @@ def switch_software_channel(plugin, channel):
+        SW_UPDATE_TIER_BETA,
+        SW_UPDATE_TIER_DEV,
+        SW_UPDATE_TIER_PROD,
+        SW_UPDATE_TIER_ALPHA,
```
Same Enum could be used here

Yelko-mrbeam commented on octoprint_mrbeam/static/js/software_channel_selector.js:
```diff
@@ -117,9 +127,18 @@ $(function () {
         };
     }
+    function reload_update_info() {
+        OctoPrint.get("plugin/mrbeam/fetch_update_info")
```
if we change to delete and also change the endpoint the change should also be reflected here

Yelko-mrbeam commented on octoprint_mrbeam/static/js/software_channel_selector.js:
```diff
@@ -117,9 +127,18 @@ $(function () {
+            .done(function (response) {
+            })
+            .fail(function () {
+                console.error("Unable to relaod update info.");
```
When would it fail? any case we should inform the user?

Yelko-mrbeam commented on octoprint_mrbeam/software_update_information.py:
```diff
@@ -32,59 +43,188 @@
+        user_notification_system.get_notification(
+            notification_id="missing_updateinformation_info", replay=False
         )
     )
```
should this be part of the else branch if no internet connection or how does this work? 😕

Yelko-mrbeam commented on setup.py:
```diff
@@ -44,10 +44,12 @@
+    "lxml == 4.6.3",
+    "numpy == 1.16.6",
+    "pyyaml == 3.10",
+    "enum == 0.4.7",
```
is this working with python 2.7? I had to use enum34==1.1.10 since something wasn't working with this... I don't remember exactly what was it. But since I added it for the MrBeamDoc and both of them provide the same package name I am not sure how will it behave... 
https://github.com/mrbeam/MrBeamDoc/pull/10/files

Yelko-mrbeam commented on tests/softwareupdate/test_cloud_config.py:
```diff
@@ -0,0 +1,668 @@
+                status_code=404,
+            )
+            tag_of_github_repo = get_tag_of_github_repo("beamos_config")
+            assert tag_of_github_repo == None
```
since you are on a TestCase you could use self.assertIsNone https://docs.python.org/3/library/unittest.html#assert-methods 😄

Yelko-mrbeam commented on tests/softwareupdate/test_cloud_config.py:
```diff
@@ -0,0 +1,668 @@
+            print("config {}".format(update_config))
+            assert (
+                update_config["octoprint"]
+                == target_octoprint_config[get_tier_by_id(tier)]
```
seems like formating is not right or is it? you could use self.assertEquals https://docs.python.org/3/library/unittest.html#assert-methods

Yelko-mrbeam commented on octoprint_mrbeam/software_update_information.py:
```diff
@@ -98,242 +238,209 @@ def switch_software_channel(plugin, channel):
-    pluginInfo = plugin._plugin_manager.get_plugin_info(module_id)
-    if pluginInfo is None:
-        display_version = None
+def _set_info_from_cloud_config(plugin, tier, beamos_date, cloud_config):
```
why is this a private method? since we are testing it I assume this should be public, I also don't see why this shouldn't be used from the outside...

JFSolution replied:
changed the testing to test the public method `get_update_information`

Yelko-mrbeam commented on tests/softwareupdate/test_cloud_config.py:
```diff
@@ -0,0 +1,668 @@
+# coding=utf-8
```
Great job on the testing!! 👏 🎉 

could we include maybe one more test for the user_notification_system to verify that is triggered when we wanted to be triggered?

Commit:
fix connection check settings

Commit:
fix connection checker settings
add user notification with errorcode
fix enum version

Commit:
fix tests, add notification test

Commit:
added check if github rate limit exceeded

Commit:
added PR feedback

Commit:
fix

Commit:
change doc strings to google formate, add missing docstrings

Commit:
move target config and mock config to seperate files and some cleanup

Commit:
refactor sw update tiers

Commit:
PR feedback
fix for software_tiers_available

Yelko-mrbeam commented on octoprint_mrbeam/software_update_information.py:
```diff
@@ -1,27 +1,42 @@
-    SW_UPDATE_TIER_DEV: "develop",
-}
+class SWUpdateTier(Enum):
+    PROD = "PROD"
```
shouldn't this be STABLE? 😕

JFSolution replied:
yes you are right

JFSolution replied:
But if I change it now I need additional migration  :sweat_smile:

JFSolution replied:
I changed the name to STABLE but kept the value

Yelko-mrbeam commented on octoprint_mrbeam/software_update_information.py:
```diff
@@ -1,27 +1,42 @@
-# add to the display name to modules that should be shown at the top of the list
-SORT_UP_PREFIX = " "
+SW_UPDATE_TIERS_DEV = [SWUpdateTier.ALPHA, SWUpdateTier.DEV]
+SW_UPDATE_TIERS_STABLE = [SWUpdateTier.PROD, SWUpdateTier.BETA]
```
and this SW_UPDATE_TIERS_PROD?

Yelko-mrbeam replied:
I really like this idea to have them grouped here and detached from the rest of the code 😄

JFSolution replied:
yes you are right

Yelko-mrbeam commented on octoprint_mrbeam/software_update_information.py:
```diff
@@ -1,27 +1,42 @@
+SW_UPDATE_TIERS_STABLE = [SWUpdateTier.PROD, SWUpdateTier.BETA]
+SW_UPDATE_TIERS = SW_UPDATE_TIERS_DEV + SW_UPDATE_TIERS_STABLE
 
+DEFAULT_REPO_BRANCH_ID = {
```
https://stackoverflow.com/questions/54793033/python-tuple-and-enum/70325042#70325042
this might be interesting for this... we could just add the branch as a property of the enum maybe, what do you think @Josef-MrBeam ?

JFSolution replied:
I checked and need to change to much to get it working and don't see to much benefit

Yelko-mrbeam commented on octoprint_mrbeam/software_update_information.py:
```diff
@@ -32,308 +47,492 @@
+        s.keep_alive = False
+
+        response = s.request("GET", url, headers=headers)
+    except requests.ReadTimeout:
```
I think here the same MaxRetryError as before should be catched right?

Yelko-mrbeam commented on octoprint_mrbeam/software_update_information.py:
```diff
@@ -32,308 +47,492 @@
 
-def get_update_information(plugin):
-    result = dict()
+def _get_tag_of_github_repo(repo):
```
This could be a public method right? 🤔

JFSolution replied:
you are right

Yelko-mrbeam commented on octoprint_mrbeam/software_update_information.py:
```diff
@@ -32,308 +47,492 @@
+        return None
+
+
+def _get_config_of_tag(tag):
```
This one also could be public I think... 🤔 or why are they private? the do have a documentation and I think they could be used from another class if needed... What speaks against it? @Josef-MrBeam ? we can leave both as they are, your call 😄 but it would be easier to create smaller unit tests just for this methods

JFSolution replied:
for this one I would keep it as it is as it will get the content of the sw-update-conf file for a given tag

Commit:
dont't fix versionnumber

Yelko-mrbeam commented on octoprint_mrbeam/software_update_information.py:
```diff
@@ -32,308 +47,492 @@
+    """
+    if tier in SW_UPDATE_TIERS:
+        # merge default settings and input is master
+        input_moduleconfig = dict_merge(defaultsettings, input_moduleconfig)
```
should we rename the assignation to something more accurate like updated_module_configuration? I am guessing the prefix input_ comes from being a parameter? or does it make sense to keep it? @Josef-MrBeam ?
`
        moduleconfig_updated = dict_merge(defaultsettings, input_moduleconfig)
`

JFSolution replied:
yes this was used as it is the input through a parameter, so I would keep it

Yelko-mrbeam commented on tests/softwareupdate/test_cloud_config.py:
```diff
@@ -0,0 +1,456 @@
+        return None
+
+
+class DummyConnectifityChecker:
```
Connectifity? 😋 is that a typo?

Yelko-mrbeam commented on tests/softwareupdate/test_cloud_config.py:
```diff
@@ -0,0 +1,456 @@
+        self.user_notification_system = user_notification_system(self)
+
+
+@ddt
```
you are not using this in the end right?

JFSolution replied:
no, I will delete it

Yelko-mrbeam commented:
everything looks good just missing one except MaxRetryError and some potential typos. Great job on the testing!

Yelko-mrbeam commented on octoprint_mrbeam/software_update_information.py:
```diff
@@ -32,308 +47,492 @@
+        plugin: Mr Beam Plugin
+
+    Returns:
+        software update informationen for the module
```
informationen -> information/informations

Commit:
fixed tier enum use
fixed tests

Commit:
added doc string for the tests

Commit:
PR suggested changes

Commit:
PR suggested changes

Yelko-mrbeam commented:
@Josef-MrBeam conflicts need to be resolved

JFSolution commented:
> @Josef-MrBeam conflicts need to be resolved

I will rebase on the current develop

Commit:
change Migration to new concept

JFSolution committed:
Merge branch 'SW-527-mr-beam-tag-based-updates' into feature/SW-653-mr-beam-to-use-update-config-from-cloud

Commit:
remove group and severity of errorcode

JFSolution commented on octoprint_mrbeam/software_update_information.py:
```diff
@@ -32,308 +47,493 @@
+            "default": {},
+            "modules": {
+                "mrbeam": {
+                    "name": " MrBeam Plugin offline2",
```
remove offline2

JFSolution commented on octoprint_mrbeam/software_update_information.py:
```diff
@@ -32,308 +47,493 @@
+        return None
+
 
+def get_update_information(plugin):
```
add try except so the plugin won't break -> log to analytics

JFSolution commented on octoprint_mrbeam/software_update_information.py:
```diff
@@ -32,308 +47,493 @@
+        sw_update_file_path = os.path.join(
+            plugin._settings.getBaseFolder("base"), SW_UPDATE_INFO_FILE_NAME
+        )
+        try:
```
check if we need to except more errors, for example for dumps

JFSolution commented on octoprint_mrbeam/software_update_information.py:
```diff
@@ -32,308 +47,493 @@
+            with open(sw_update_file_path, "w") as f:
+                f.write(json.dumps(sw_update_config))
+        except IOError:
+            plugin._logger.error("can't create update info file")
```
add this to analytics

JFSolution replied:
@ahmed-mrbeam errors will already be added to analytics

JFSolution commented on octoprint_mrbeam/software_update_information.py:
```diff
@@ -32,308 +47,493 @@
+    import json
+
+    try:
+        url = "https://api.github.com/repos/mrbeam/" + repo + "/tags"
```
change to `"https://api.github.com/repos/mrbeam/{repo}/tags".format(repo=repo)`

JFSolution commented on octoprint_mrbeam/static/js/software_channel_selector.js:
```diff
@@ -117,9 +127,18 @@ $(function () {
+            .done(function (response) {
+            })
+            .fail(function () {
+                console.error("Unable to reload update info.");
```
add analytics

JFSolution commented on setup.py:
```diff
@@ -44,10 +44,12 @@
-    "pillow",
-    "lxml",
-    "numpy",
+    "webcolors <= 1.10",
```
move this definition of the versions to a seperate issue

JFSolution commented on setup.py:
```diff
@@ -82,6 +84,7 @@
             "files/migrate_logrotate/*",
             "files/material_settings/*",
             "files/camera/*",
+            "files/software_update/*",
```
this is not needed anymore, please remove

JFSolution commented on octoprint_mrbeam/migrate.py:
```diff
@@ -49,6 +52,7 @@ class Migration(object):
     VERSION_UPDATE_CUSTOM_MATERIAL_SETTINGS = "0.9.9"
     VERSION_UPDATE_OCTOPRINT_PRERELEASE_FIX = "0.9.10"
     VERSION_UPDATE_FORCE_FOCUS_REMINDER = "0.10.0"
+    VERSION_ENABLE_ONLINECHECK = "0.11.0"
```
remove

JFSolution commented on octoprint_mrbeam/migrate.py:
```diff
@@ -404,6 +412,15 @@ def save_current_version(self):
+
+    @restart.setter
+    def restart(self, value):
+        if self._restart == 0 or value < self._restart:
```
change so None won't override the restart

JFSolution commented on octoprint_mrbeam/__version.py:
```diff
@@ -1 +1 @@
-__version__ = "0.10.3"
+__version__ = "0.11.0"
```
revert

Commit:
added feeback from PR:
- changed logging string to use format
- added analytics
- reverted version bump
- removed unused stuff
- reverted required packages version fix

Commit:
convert docstring to google docstring
